### PR TITLE
#635 move authentication to a background thread

### PR DESCRIPTION
**Pull request by jsoberg**
Ok, so I added an ASyncTask to run the authentication code, with a callback to the Activity when it completes (still finishing the Activity on the main thread). While the task is running, I added an indeterminate progress bar that replaces the "unlock" button to provide some feedback to the user: 

https://user-images.githubusercontent.com/1057406/106338658-56f2c580-6262-11eb-8816-9c88006b6d2f.mp4

One small caveat here is that I call `.get()` on the task if the Activity is being paused. I did this to prevent the task from calling back to a dead activity for the case where the user initiates the unlock process and rotates their phone (or some other config change). 

If this situation happens, the task will complete and the Activity will finish as expected, however the main thread will be locked while `.get()` is called (as it was before). Locking the main thread here isn't ideal, but seeing as this is a short-lived (1-2 second) task and the situation is unlikely to happen it's probably not a big issue. 

To get around this we'd probably have to start using a retained fragment within the Activity (to store the task), and then in `onCreate` check for that retained fragment and tie the newly created Activity into it for a callback. This is definitely doable, however it'd probably make more sense to wait and see if we implement a general background execution scheme for the app before doing that work.
    
Let me know if you see any issues and I'll make necessary changes. I tested this on a Samsung device and on an emulator for both the cases of regular login and changing/upgrading authentication, with configuration changes in between and didn't run into any issues with the changes.

### CODE REVIEW

**Comment by flocke** [comment]
Thank you a lot, I hope I find the time to review this soon.

**Comment by flocke** [comment]
I haven't found the time yet to review the PR in detail, but I noticed you used an AsyncTask. I looked into that as well for the creating backups in a background process and noticed that it is set to be deprecated in API level 30: https://developer.android.com/reference/android/os/AsyncTask

Do you think it would be feasible to use something else to do this? I decided on using a native Java Runnable and an ExecutorService. You can take look at it in this commit: https://github.com/andOTP/andOTP/commit/feeeba7294864d10f474aae093f73bf72e53a810

If you want to put in the extra work I would appreciate it if you would move it of AsyncTask yourself. But I can understand if you don't have the time or don't want to. In that case I will merge it as it is (after a more detailed review) and port it myself later on when I find some time.

**Comment by jsoberg** [comment]
Oh sorry, I completely forgot that ASyncTask was being deprecated. I can definitely move it into using an executor with a main thread Handler instead. We could also pull in RxJava (though that's adding another dependency), but it has some nice threading utilities built in (default schedulers for background work, Android main thread scheduler for pushing to the main thread, etc). Let me know if you prefer one over the other, it's not a big change either way so if I don't hear back I'll just use an executor.

**Comment by flocke** [comment]
I would prefer to avoid additional dependencies if not strictly necessary, so I would prefer an executor.

Even though I posted my commit as example, you probably should not look at it to much. It was my first time doing something like this in an Android app, so it's provably not very clean. You seem to have some more experience in that regard, so I will probably look at your code afterwards trying to learn something from it ;)

**Comment by jsoberg** [comment]
That makes sense, I'll stick with an executor service. Haha I have some experience with threading in Android, but any way you go it's never fun to tie background work back to the UI in a clean way (hence my .get() call in the code I put up). We might be able to use a base executor and send a lambda/callable in to do background work that'll make things like this easier in the future, I'll definitely take a look while modifying my code within the next few days

**Comment by jsoberg** [comment]
Added a base `UiBackedBackgroundTask` class that can hopefully be reused for similar instances where we want to run a background task that calls back to a UI on the main thread with a result, and converted the `AuthenticationTask` to extend this instead of ASyncTask.

I went ahead and implemented the retained fragment concept that I was talking about earlier while I was at it. This means that on configuration changes (i.e the user rotates their device, goes home, etc), the task will continue to run and if the activity is eventually brought back, we can reconnect to the task and get our callback. I made sure to set the callback to null when pausing so that we wouldn't ever get a callback to a dead activity instance. 

All in all, the user can rotate their device and the main thread won't be paused (it will just show progress again when recreated, until the task completes). If the task completes between configuration changes, it retains the result and just calls back immediately when a new callback is set (in the activity's onResume).

File: app/src/main/java/org/shadowice/flocke/andotp/Activities/AuthenticateActivity.java
Comment (jsoberg) [inline app/src/main/java/org/shadowice/flocke/andotp/Activities/AuthenticateActivity.java:249]: Just noticed this weird double javadoc link, I'll fix that real quick on my lunch break

**Comment by flocke** [comment]
I finally found the time to review the code and I didn't really find anything I would change in the implementation.

The only problem I found come from the "Re-lock on screen off" and "Re-lock when going into the background" options.
With your retained fragment approach the unlocking continues in both cases when coming back from the background or unlocking the screen again. I would prefer the unlocking to be canceled in that case so we don't return to an unlocked activity, which would contradict the settings in my opinion.

**Comment by jsoberg** [comment]
> 
> 
> I finally found the time to review the code and I didn't really find anything I would change in the implementation.
> 
> The only problem I found come from the "Re-lock on screen off" and "Re-lock when going into the background" options.
> With your retained fragment approach the unlocking continues in both cases when coming back from the background or unlocking the screen again. I would prefer the unlocking to be canceled in that case so we don't return to an unlocked activity, which would contradict the settings in my opinion. But only if one of the settings is enabled and the corresponding event (screen off or going into background) happens during the unlock task.
> 
> Edit: And I'm definitely going to refactor my backup task to use your `UiBackedBackgroundTask` and the retained fragment approach once this merged.

Good point, thank you for catching that. I'll make sure to check that setting for this case and cancel the task if it applies. I'm not sure off the top of my head how to tell when we're entering/coming back from a lock or background event vs a normal config change, but I'll take a look where that setting is being checked currently and add that in.

**Comment by jsoberg** [comment]
I added the ability to cancel a `UiBackedBackgroundTask`, and then used the `MainActivity` as a reference for cancelling the task on screen off/background if set to do so in the settings. I tested this on my device and it seemed to work fine (though in some cases it seems like Android kills the Activity on screen off, which ends with the same result anyway), but let me know if you see any issues with it.

File: app/src/main/java/org/shadowice/flocke/andotp/Activities/AuthenticateActivity.java
Comment (flocke) [inline app/src/main/java/org/shadowice/flocke/andotp/Activities/AuthenticateActivity.java:196]: Could you do something like this here instead (not sure if this is the best way but it works):
```
if (taskFragment != null) {
    if (taskFragment.task.isCanceled()) {
        passwordInput.setText("");
        passwordInput.requestFocus();

        InputMethodManager keyboard = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        keyboard.showSoftInput(passwordInput, 0);
    } else {
        taskFragment.task.setCallback(this::handleResult);
        setupUiForTaskState(true);
    }
}
```

Otherwise the password stays entered when returning from background, which makes the re-lock useless.

And could exclude the `AuthenticateActivity` from being finished in `destroyIfNotMain()` in the `BaseActivity` as well? Otherwise it will show the "Authentication failed, closing" warning after screen off and end the app.

The manual `showSoftInput` seems to be required after screen-off, otherwise it did not show the keyboard again for me.

**Review by flocke — CHANGES_REQUESTED** [review]
I have one small change request to handle the re-lock cases a little better, after that I think this is ready to be merged.

**Comment by jsoberg** [comment]
Thanks for the review, I'll make sure to add all of that in. I noticed the failure toast when the Activity dies and that definitely makes sense to not show when it's killed. Did you think we should also prevent that toast from showing when finishing from onBackPressed()?

**Comment by jsoberg** [comment]
Added the requested changes:

- Modified BaseActivity to allow for destroyOnScreenOff/destroyIfNotMain functionality to be overridden. I thought Android was just destroying the Activity on screen off here, sorry about that.
- Made sure to reset the password input and request focus when coming back and seeing our task has been canceled. I made sure to remove the task fragment (when we're able) as well after seeing a canceled task, since otherwise the password input would be reset on each configuration change after canceling a task.

Tested those changes and everything seems to work as expected, but let me know if you see anything else and I'll be happy to change it

**Review by flocke — APPROVED** [review]

**Comment by flocke** [comment]
Looks good, thank you for implementing this!

**Pull request merged**
